src/core/heavyGraph.js

```diff
@@ -0,0 +1,122 @@
+// @flow
+import {
+  type Edge,
+  type Node,
+  type EdgeAddressT,
+  type NodeAddressT,
+  Graph,
+} from "./graph";
+import {
+  type NodeWeight,
+  type EdgeWeight,
+  type WeightsComparison,
+} from "./weights";
+
+/**
+ * An entry in the array returned by AddressModule.toParts()
+ */
+export type AddressPart = string;
+
+/**
+ * the core structure in the Node weight-resolution tree
+ */
+export type NodeWeightBranch = {
+  +weight?: NodeWeight,
+  +children: Map<AddressPart, NodeWeightBranch>,
+};
+
+/**
+ * the core structure in the Edge weight-resolution tree
+ */
+export type EdgeWeightBranch = {
+  +weight?: EdgeWeight,
+  +children: Map<AddressPart, EdgeWeightBranch>,
+};
+
+/**
+ * The HeavyGraph class is a replacement for the WeightedGraph type. This class
+ * eliminates the need give every node and edge that exists a weight. Rather,
+ * address prefixes can be assigned a weight and when a node or edge weight is
+ * queried, the longest prefix with a weight assigned will have its value
+ * returned.
+ *
+ * Over the longer-term, WeightedGraph should be
+ * refactored out of existence HeavyGraph should completely replace it.
+ */
+export class HeavyGraph {
```

Comment:
Let's call it `WeightedGraph2`, with an intent to rename it to `WeightedGraph` once we've updated every user. I think the name "WeightedGraph" is much more descriptive and intuitive than "HeavyGraph".